compiler/resolution/interfaceResolution.cpp

```diff
@@ -625,6 +626,100 @@ static bool resolveAssociatedTypes(InterfaceSymbol* isym, ImplementsStmt* istm,
   return atSuccess;
 }
 
+// instantiateAggregateTypes
+
+// Returns a type if it is a replacement instantiation, NULL otherwise.
+static Type* instantiateOneAggregateType(SymbolMap &fml2act,
+                                         BlockStmt*  holder,
+                                         Type*        origT) {
+  if (Symbol* replacement = fml2act.get(origT->symbol))
+    return toTypeSymbol(replacement)->type; // already handled
+
+  AggregateType* at = toAggregateType(origT->getValType());
+  // CG TODO: also handle DecoratedClassType
+  if (at == nullptr) return nullptr; // there will be nothing interesting
+
+  AggregateType* atgen = at->instantiatedFrom;
+  if (atgen == nullptr) return nullptr; // 'at' is not a generic type
+  // CG TODO: can 'atgen' also be an instantiated generic with ifc types?
+
+  // We may or may not use its contents.
+  std::vector<Type*> subs;  // used only when there are new substitution(s)
+  bool gotSub = false;
+  form_Map(SymbolMapElem, elem, at->substitutions) {
+    if (elem->key->hasFlag(FLAG_PARAM)) continue;
+    Type* srcT = elem->value->type;
+    Type* inst = instantiateOneAggregateType(fml2act, holder, srcT);
+    subs.push_back(inst != nullptr ? inst : srcT);
+    if (inst != nullptr) gotSub = true;
+  }
+
+  if (! gotSub) return nullptr; // nothing to replace
+
+  // We got something to handle. Call atgen->generateType().
+  CallExpr* genCall = new CallExpr(atgen->symbol);
+  holder->insertAtHead(genCall);
+  int i = 0;
+  form_Map(SymbolMapElem, elem, at->substitutions) {
+    Symbol* instSym = nullptr;
+    // follow addArgForGenericField()
+    if (elem->key->hasFlag(FLAG_PARAM)) // param field
+      instSym = elem->value; // apply the same substitution as for 'at'
+    else // var or type field
+      instSym = subs[i++]->symbol;
+    genCall->insertAtTail(new NamedExpr(elem->key->name, new SymExpr(instSym)));
+  }
+
+  AggregateType* instT = atgen->generateType(genCall, "<internal error>");
+  genCall->remove();
+  fml2act.put(at->symbol, instT->symbol);
+
+  Type* result = instT;
+  if (at != origT) {
+    // We started with a ref type, make one here, too.
+    makeRefType(instT);
+    result = instT->refType;
+    fml2act.put(origT->symbol, result->symbol);
+  }
+
+  return result;
+}
+
+//
+// Extends fml2act to map each aggregate type over interface type(s),
```

Comment:
Can we say something in this comment like:

```
Maps any uses of "Self" within an interface declaration to a new aggregate type
with "Self" replaced according to fml2act.
...
```

Or something along those lines? This may not be that important since implementation is ongoing, but I found it hard to grok.